### ISSUE TITLE
subprocess.Popen Hangs

### ISSUE THREAD

**Issue by gnossen** [grpc/grpc#17986]
### What version of gRPC and what language are you using?
 master
 
### What operating system (Linux, Windows, …) and version?
 Linux, Debian 4.18.10
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
Python 3.5.4
 
### What did you do?
At the suggestion of @GuillaumeCisco in #17203, I ran the following client -- a simple modification of the `greeter_client.py` example:

```python
from __future__ import print_function
import logging

import grpc

import hellostreamstreamworld_pb2
import hellostreamstreamworld_pb2_grpc

import subprocess
import sys


def run():
    with grpc.insecure_channel('localhost:50051') as channel:
        stub = hellostreamstreamworld_pb2_grpc.MultiGreeterStub(channel)

        def get_request():
            for i in range(9999999):
                p = subprocess.Popen(["/bin/echo", "Hello", "world"])
                p.wait()
                print("Echo return code: {}".format(p.returncode))
                yield hellostreamstreamworld_pb2.HelloRequest(name=str(i))

        responses = stub.sayHello(get_request())
        for response in responses:
            print("Greeter client received: " + response.message)


if __name__ == '__main__':
    logging.basicConfig()
    run()
```
 
### What did you expect to see?
The client should continue for 9999999 iterations, printing "Hello world" each iteration.
 
### What did you see instead?
 
 Instead, the process nondeterministically hangs after a small number of iterations in the `subprocess.Popen` call. I've further tracked the hang to a call to [`subprocess_fork_exec`](https://github.com/python/cpython/blob/6f9bc72c79c3262e5d0f2c0e96b016477399cfb1/Modules/_posixsubprocess.c#L557). There appears to be some sort of lock contention happening here.

The backtrace doesn't seem particularly helpful:
```
Thread 6 (Thread 0x7ffff1f98700 (LWP 132970)):
#0  pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
#1  0x00007ffff64f5f72 in gpr_cv_wait (cv=0x555555fd48e0, mu=0x555555fd48b8, abs_deadline=...) at src/core/lib/gpr/sync_posix.cc:172
#2  0x00007ffff64f76c3 in grpc_core::internal::ExecCtxState::IncExecCtxCount (this=<optimized out>) at src/core/lib/gprpp/fork.cc:67
#3  grpc_core::Fork::IncExecCtxCount () at src/core/lib/gprpp/fork.cc:215
#4  0x00007ffff6514978 in grpc_core::ApplicationCallbackExecCtx::ApplicationCallbackExecCtx (this=0x7ffff1f96bf0) at ./src/core/lib/iomgr/exec_ctx.h:235
#5  grpc_core::Executor::RunClosures (executor_name=0x7ffff671713a "default-executor", list=...) at src/core/lib/iomgr/executor.cc:119
#6  0x00007ffff65156f7 in grpc_core::Executor::SetThreading (this=0x555555f77ab0, threading=<optimized out>) at src/core/lib/iomgr/executor.cc:203
#7  0x00007ffff651617c in grpc_core::Executor::SetThreadingAll (enable=false) at src/core/lib/iomgr/executor.cc:458
#8  0x00007ffff65165ee in grpc_prefork () at src/core/lib/iomgr/fork_posix.cc:74
#9  0x00007ffff6d7e2f5 in __libc_fork () at ../sysdeps/nptl/fork.c:96
#10 0x000055555578cb5e in subprocess_fork_exec (self=<optimized out>, args=<optimized out>) at ../Modules/_posixsubprocess.c:676
#11 0x00005555557112ff in PyCFunction_Call (func=0x7ffff573b828, args=<optimized out>, kwds=<optimized out>) at ../Objects/methodobject.c:109
#12 0x00005555556d1ede in call_function (oparg=<optimized out>, pp_stack=0x7ffff1f97070) at ../Python/ceval.c:4758
#13 PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at ../Python/ceval.c:3289
#14 0x00005555556da5e7 in _PyEval_EvalCodeWithName (_co=0x7ffff5752d20, globals=<optimized out>, locals=<optimized out>, args=<optimized out>, argcount=<optimized out>, kws=0x7fffe0001488, kwcount=0, defs=0x0, defcount=0, kwdefs=0x0, closure=0x0, name=<optimized out>, 
    qualname=0x7ffff5753420) at ../Python/ceval.c:4071
#15 0x00005555556d6634 in fast_function (nk=0, na=19, n=<optimized out>, pp_stack=0x7ffff1f97270, func=<optimized out>) at ../Python/ceval.c:4866
#16 call_function (oparg=<optimized out>, pp_stack=0x7ffff1f97270) at ../Python/ceval.c:4783
#17 PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at ../Python/ceval.c:3289
#18 0x00005555556dae0d in _PyEval_EvalCodeWithName (qualname=0x0, name=<optimized out>, closure=<optimized out>, kwdefs=<optimized out>, defcount=<optimized out>, defs=<optimized out>, kwcount=<optimized out>, kws=<optimized out>, argcount=<optimized out>, args=<optimized out>, 
    locals=<optimized out>, globals=<optimized out>, _co=<optimized out>) at ../Python/ceval.c:4071
#19 PyEval_EvalCodeEx () at ../Python/ceval.c:4092
#20 0x0000555555713208 in function_call.lto_priv.388 (func=func@entry=0x7ffff52c6d08, arg=0x7ffff4898788, kw=0x0) at ../Objects/funcobject.c:627
#21 0x000055555575ac47 in PyObject_Call (func=0x7ffff52c6d08, arg=<optimized out>, kw=<optimized out>) at ../Objects/abstract.c:2166
#22 0x000055555567da1e in method_call.lto_priv.302 (func=0x7ffff52c6d08, func@entry=0x7ffff5285e88, arg=0x7ffff4898788, kw=0x0) at ../Objects/classobject.c:330
#23 0x000055555575ac47 in PyObject_Call (func=0x7ffff5285e88, arg=<optimized out>, kw=<optimized out>) at ../Objects/abstract.c:2166
#24 0x0000555555709ac9 in slot_tp_init (self=self@entry=0x7ffff40996a0, args=args@entry=0x7ffff4099780, kwds=kwds@entry=0x0) at ../Objects/typeobject.c:6314
#25 0x0000555555718924 in type_call.lto_priv.74 (type=<optimized out>, type@entry=0x555555dd26d8, args=0x7ffff4099780, kwds=0x0) at ../Objects/typeobject.c:905
#26 0x000055555575ac47 in PyObject_Call (func=0x555555dd26d8, arg=<optimized out>, kw=<optimized out>) at ../Objects/abstract.c:2166
#27 0x00005555556d61b3 in do_call (nk=<optimized out>, na=<optimized out>, pp_stack=0x7ffff1f97620, func=<optimized out>) at ../Python/ceval.c:4989
#28 call_function (oparg=<optimized out>, pp_stack=0x7ffff1f97620) at ../Python/ceval.c:4785
#29 PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at ../Python/ceval.c:3289
#30 0x00005555556dcf29 in gen_send_ex.lto_priv.1931 (closing=0, exc=<optimized out>, arg=0x0, gen=0x7ffff406ff68) at ../Objects/genobject.c:141
#31 gen_iternext.lto_priv.379 (gen=0x7ffff406ff68) at ../Objects/genobject.c:454
#32 builtin_next (self=<optimized out>, args=<optimized out>) at ../Python/bltinmodule.c:1286
#33 0x00005555557112ff in PyCFunction_Call (func=0x7ffff6cbe6c0, args=<optimized out>, kwds=<optimized out>) at ../Objects/methodobject.c:109
#34 0x00005555556d1ede in call_function (oparg=<optimized out>, pp_stack=0x7ffff1f977e0) at ../Python/ceval.c:4758
#35 PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at ../Python/ceval.c:3289
#36 0x00005555556db74f in _PyEval_EvalCodeWithName (qualname=0x0, name=<optimized out>, closure=<optimized out>, kwdefs=<optimized out>, defcount=<optimized out>, defs=<optimized out>, kwcount=<optimized out>, kws=<optimized out>, argcount=<optimized out>, args=<optimized out>, 
    locals=<optimized out>, globals=<optimized out>, _co=<optimized out>) at ../Python/ceval.c:4071
#37 PyEval_EvalCodeEx () at ../Python/ceval.c:4092
#38 0x00005555557132a2 in function_call.lto_priv.388 (func=func@entry=0x7ffff409d9d8, arg=0x7ffff6c86048, kw=0x7ffff409b548) at ../Objects/funcobject.c:627
#39 0x000055555575ac47 in PyObject_Call (func=0x7ffff409d9d8, arg=<optimized out>, kw=<optimized out>) at ../Objects/abstract.c:2166
#40 0x00005555556d3b4e in ext_do_call (nk=<optimized out>, na=<optimized out>, flags=<optimized out>, pp_stack=0x7ffff1f97a78, func=0x7ffff409d9d8) at ../Python/ceval.c:5087
#41 PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at ../Python/ceval.c:3328
#42 0x00005555556d5acf in fast_function (nk=<optimized out>, na=<optimized out>, n=<optimized out>, pp_stack=0x7ffff1f97b70, func=<optimized out>) at ../Python/ceval.c:4856
#43 call_function (oparg=<optimized out>, pp_stack=0x7ffff1f97b70) at ../Python/ceval.c:4783
#44 PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at ../Python/ceval.c:3289
#45 0x00005555556d5acf in fast_function (nk=<optimized out>, na=<optimized out>, n=<optimized out>, pp_stack=0x7ffff1f97c70, func=<optimized out>) at ../Python/ceval.c:4856
#46 call_function (oparg=<optimized out>, pp_stack=0x7ffff1f97c70) at ../Python/ceval.c:4783
#47 PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at ../Python/ceval.c:3289
#48 0x00005555556dae0d in _PyEval_EvalCodeWithName (qualname=0x0, name=<optimized out>, closure=<optimized out>, kwdefs=<optimized out>, defcount=<optimized out>, defs=<optimized out>, kwcount=<optimized out>, kws=<optimized out>, argcount=<optimized out>, args=<optimized out>, 
    locals=<optimized out>, globals=<optimized out>, _co=<optimized out>) at ../Python/ceval.c:4071
#49 PyEval_EvalCodeEx () at ../Python/ceval.c:4092
#50 0x0000555555713193 in function_call.lto_priv.388 (func=func@entry=0x7ffff6a50ea0, arg=0x7ffff408ac88, kw=0x0) at ../Objects/funcobject.c:627
#51 0x000055555575ac47 in PyObject_Call (func=0x7ffff6a50ea0, arg=<optimized out>, kw=<optimized out>) at ../Objects/abstract.c:2166
#52 0x000055555567da1e in method_call.lto_priv.302 (func=0x7ffff6a50ea0, func@entry=0x7ffff6c38f88, arg=0x7ffff408ac88, kw=0x0) at ../Objects/classobject.c:330
#53 0x000055555575ac47 in PyObject_Call (func=0x7ffff6c38f88, arg=<optimized out>, kw=<optimized out>) at ../Objects/abstract.c:2166
#54 0x00005555556d0500 in PyEval_CallObjectWithKeywords (func=0x7ffff6c38f88, arg=0x7ffff6c86048, kw=<optimized out>) at ../Python/ceval.c:4633
#55 0x00005555557a1c12 in t_bootstrap (boot_raw=0x55555601ed80) at ../Modules/_threadmodule.c:998
#56 0x00007ffff7bc3494 in start_thread (arg=0x7ffff1f98700) at pthread_create.c:333
#57 0x00007ffff6daea8f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97

Thread 5 (Thread 0x7ffff2799700 (LWP 132969)):
#0  pthread_cond_timedwait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_timedwait.S:225
#1  0x00005555556d156d in PyCOND_TIMEDWAIT (cond=0x555555bd6aa0 <gil_cond>, mut=0x555555bd6a60 <gil_mutex>, us=5000) at ../Python/condvar.h:103
#2  take_gil (tstate=0x55555601eca0) at ../Python/ceval_gil.h:224
#3  0x00005555556d1993 in PyEval_RestoreThread (tstate=0x55555601eca0) at ../Python/ceval.c:450
#4  0x00007ffff647b526 in __pyx_f_4grpc_7_cython_6cygrpc__next (__pyx_v_c_completion_queue=0x55555601c3f0, __pyx_v_deadline=<optimized out>) at src/python/grpcio/grpc/_cython/cygrpc.cpp:24149
#5  0x00007ffff64d2c70 in __pyx_f_4grpc_7_cython_6cygrpc__latent_event (__pyx_v_deadline=0x555555bca6a0 <_Py_NoneStruct>, __pyx_v_c_completion_queue=<optimized out>) at src/python/grpcio/grpc/_cython/cygrpc.cpp:24405
#6  __pyx_f_4grpc_7_cython_6cygrpc__next_call_event (__pyx_v_c_completion_queue=<optimized out>, __pyx_v_on_success=__pyx_v_on_success@entry=0x7ffff40688e8, __pyx_v_deadline=__pyx_v_deadline@entry=0x555555bca6a0 <_Py_NoneStruct>, __pyx_v_channel_state=0x7ffff40887b8)
    at src/python/grpcio/grpc/_cython/cygrpc.cpp:11616
--Type <RET> for more, q to quit, c to continue without paging-- 
#7  0x00007ffff64d44c9 in __pyx_pf_4grpc_7_cython_6cygrpc_7Channel_6next_call_event (__pyx_v_self=<optimized out>) at src/python/grpcio/grpc/_cython/cygrpc.cpp:17319
#8  __pyx_pw_4grpc_7_cython_6cygrpc_7Channel_7next_call_event (__pyx_v_self=<optimized out>, unused=<optimized out>) at src/python/grpcio/grpc/_cython/cygrpc.cpp:17072
#9  0x00005555556d64d8 in call_function (oparg=<optimized out>, pp_stack=0x7ffff27987e0) at ../Python/ceval.c:4737
#10 PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at ../Python/ceval.c:3289
#11 0x00005555556db2b2 in _PyEval_EvalCodeWithName (qualname=0x0, name=<optimized out>, closure=<optimized out>, kwdefs=<optimized out>, defcount=<optimized out>, defs=<optimized out>, kwcount=<optimized out>, kws=<optimized out>, argcount=<optimized out>, args=<optimized out>, 
    locals=<optimized out>, globals=<optimized out>, _co=<optimized out>) at ../Python/ceval.c:4071
#12 PyEval_EvalCodeEx () at ../Python/ceval.c:4092
#13 0x00005555557132a2 in function_call.lto_priv.388 (func=func@entry=0x7ffff409d950, arg=0x7ffff6c86048, kw=0x7ffff409b408) at ../Objects/funcobject.c:627
#14 0x000055555575ac47 in PyObject_Call (func=0x7ffff409d950, arg=<optimized out>, kw=<optimized out>) at ../Objects/abstract.c:2166
#15 0x00005555556d3b4e in ext_do_call (nk=<optimized out>, na=<optimized out>, flags=<optimized out>, pp_stack=0x7ffff2798a78, func=0x7ffff409d950) at ../Python/ceval.c:5087
#16 PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at ../Python/ceval.c:3328
#17 0x00005555556d5acf in fast_function (nk=<optimized out>, na=<optimized out>, n=<optimized out>, pp_stack=0x7ffff2798b70, func=<optimized out>) at ../Python/ceval.c:4856
#18 call_function (oparg=<optimized out>, pp_stack=0x7ffff2798b70) at ../Python/ceval.c:4783
#19 PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at ../Python/ceval.c:3289
#20 0x00005555556d5acf in fast_function (nk=<optimized out>, na=<optimized out>, n=<optimized out>, pp_stack=0x7ffff2798c70, func=<optimized out>) at ../Python/ceval.c:4856
#21 call_function (oparg=<optimized out>, pp_stack=0x7ffff2798c70) at ../Python/ceval.c:4783
#22 PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at ../Python/ceval.c:3289
#23 0x00005555556dae0d in _PyEval_EvalCodeWithName (qualname=0x0, name=<optimized out>, closure=<optimized out>, kwdefs=<optimized out>, defcount=<optimized out>, defs=<optimized out>, kwcount=<optimized out>, kws=<optimized out>, argcount=<optimized out>, args=<optimized out>, 
    locals=<optimized out>, globals=<optimized out>, _co=<optimized out>) at ../Python/ceval.c:4071
#24 PyEval_EvalCodeEx () at ../Python/ceval.c:4092
#25 0x0000555555713193 in function_call.lto_priv.388 (func=func@entry=0x7ffff6a50ea0, arg=0x7ffff408a828, kw=0x0) at ../Objects/funcobject.c:627
#26 0x000055555575ac47 in PyObject_Call (func=0x7ffff6a50ea0, arg=<optimized out>, kw=<optimized out>) at ../Objects/abstract.c:2166
#27 0x000055555567da1e in method_call.lto_priv.302 (func=0x7ffff6a50ea0, func@entry=0x7ffff6bc96c8, arg=0x7ffff408a828, kw=0x0) at ../Objects/classobject.c:330
#28 0x000055555575ac47 in PyObject_Call (func=0x7ffff6bc96c8, arg=<optimized out>, kw=<optimized out>) at ../Objects/abstract.c:2166
#29 0x00005555556d0500 in PyEval_CallObjectWithKeywords (func=0x7ffff6bc96c8, arg=0x7ffff6c86048, kw=<optimized out>) at ../Python/ceval.c:4633
#30 0x00005555557a1c12 in t_bootstrap (boot_raw=0x555555fccdb0) at ../Modules/_threadmodule.c:998
#31 0x00007ffff7bc3494 in start_thread (arg=0x7ffff2799700) at pthread_create.c:333
#32 0x00007ffff6daea8f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97

Thread 3 (Thread 0x7ffff379b700 (LWP 132967)):
#0  pthread_cond_wait@@GLIBC_2.3.2 () at ../sysdeps/unix/sysv/linux/x86_64/pthread_cond_wait.S:185
#1  0x00007ffff64f5f72 in gpr_cv_wait (cv=0x555556019628, mu=0x5555560195f0, abs_deadline=...) at src/core/lib/gpr/sync_posix.cc:172
#2  0x00007ffff6514b2f in grpc_core::Executor::ThreadMain (arg=0x5555560195f0) at src/core/lib/iomgr/executor.cc:231
#3  0x00007ffff64f79dc in grpc_core::(anonymous namespace)::ThreadInternalsPosix::__lambda0::operator() (__closure=0x0, v=<optimized out>) at src/core/lib/gprpp/thd_posix.cc:100
#4  grpc_core::(anonymous namespace)::ThreadInternalsPosix::__lambda0::_FUN (v=<optimized out>) at src/core/lib/gprpp/thd_posix.cc:103
#5  0x00007ffff7bc3494 in start_thread (arg=0x7ffff379b700) at pthread_create.c:333
#6  0x00007ffff6daea8f in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97

Thread 1 (Thread 0x7ffff7fd1700 (LWP 132961)):
#0  0x00007ffff7bcb536 in futex_abstimed_wait_cancelable (private=0, abstime=0x0, expected=0, futex_word=0x555555fc96a0) at ../sysdeps/unix/sysv/linux/futex-internal.h:205
#1  do_futex_wait (sem=sem@entry=0x555555fc96a0, abstime=0x0) at sem_waitcommon.c:111
#2  0x00007ffff7bcb5e4 in __new_sem_wait_slow (sem=0x555555fc96a0, abstime=0x0) at sem_waitcommon.c:181
#3  0x00005555556b4475 in PyThread_acquire_lock_timed (lock=0x555555fc96a0, microseconds=-1000000, intr_flag=1) at ../Python/thread_pthread.h:352
#4  0x00005555556b81f0 in acquire_timed (timeout=<optimized out>, lock=0x555555fc96a0) at ../Modules/_threadmodule.c:68
#5  lock_PyThread_acquire_lock (self=0x7ffff4081a80, args=<optimized out>, kwds=<optimized out>) at ../Modules/_threadmodule.c:151
#6  0x0000555555711327 in PyCFunction_Call (func=0x7ffff5be53a8, args=<optimized out>, kwds=<optimized out>) at ../Objects/methodobject.c:98
#7  0x00005555556d1ede in call_function (oparg=<optimized out>, pp_stack=0x7fffffffd110) at ../Python/ceval.c:4758
#8  PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at ../Python/ceval.c:3289
#9  0x00005555556d9fc6 in _PyEval_EvalCodeWithName (_co=0x7ffff6ab69c0, globals=<optimized out>, locals=<optimized out>, args=<optimized out>, argcount=<optimized out>, kws=0x5555560203f0, kwcount=0, defs=0x7ffff6a4d6b8, defcount=1, kwdefs=0x0, closure=0x0, name=<optimized out>, 
    qualname=0x7ffff6ab91f0) at ../Python/ceval.c:4071
#10 0x00005555556d63cb in fast_function (nk=0, na=1, n=<optimized out>, pp_stack=0x7fffffffd310, func=<optimized out>) at ../Python/ceval.c:4866
#11 call_function (oparg=<optimized out>, pp_stack=0x7fffffffd310) at ../Python/ceval.c:4783
#12 PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at ../Python/ceval.c:3289
#13 0x00005555556d5acf in fast_function (nk=<optimized out>, na=<optimized out>, n=<optimized out>, pp_stack=0x7fffffffd410, func=<optimized out>) at ../Python/ceval.c:4856
#14 call_function (oparg=<optimized out>, pp_stack=0x7fffffffd410) at ../Python/ceval.c:4783
#15 PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at ../Python/ceval.c:3289
#16 0x00005555556dae0d in _PyEval_EvalCodeWithName (qualname=0x0, name=<optimized out>, closure=<optimized out>, kwdefs=<optimized out>, defcount=<optimized out>, defs=<optimized out>, kwcount=<optimized out>, kws=<optimized out>, argcount=<optimized out>, args=<optimized out>, 
    locals=<optimized out>, globals=<optimized out>, _co=<optimized out>) at ../Python/ceval.c:4071
#17 PyEval_EvalCodeEx () at ../Python/ceval.c:4092
#18 0x0000555555713193 in function_call.lto_priv.388 (func=func@entry=0x7ffff4095c80, arg=0x7ffff6a5b128, kw=0x0) at ../Objects/funcobject.c:627
#19 0x000055555575ac47 in PyObject_Call (func=0x7ffff4095c80, arg=<optimized out>, kw=<optimized out>) at ../Objects/abstract.c:2166
#20 0x000055555567da1e in method_call.lto_priv.302 (func=0x7ffff4095c80, func@entry=0x7ffff6c75988, arg=0x7ffff6a5b128, kw=0x0) at ../Objects/classobject.c:330
#21 0x000055555575ac47 in PyObject_Call (func=0x7ffff6c75988, arg=<optimized out>, kw=<optimized out>) at ../Objects/abstract.c:2166
#22 0x000055555571c15d in call_method.lto_priv.2025 (o=0x7ffff4092468, nameid=0x555555b9d990 <PyId___next__.14121>, format=<optimized out>) at ../Objects/typeobject.c:1442
#23 0x00005555556d2288 in PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at ../Python/ceval.c:3066
#24 0x00005555556d5acf in fast_function (nk=<optimized out>, na=<optimized out>, n=<optimized out>, pp_stack=0x7fffffffd930, func=<optimized out>) at ../Python/ceval.c:4856
#25 call_function (oparg=<optimized out>, pp_stack=0x7fffffffd930) at ../Python/ceval.c:4783
#26 PyEval_EvalFrameEx (f=<optimized out>, throwflag=<optimized out>) at ../Python/ceval.c:3289
#27 0x00005555556d9fc6 in _PyEval_EvalCodeWithName (_co=0x7ffff6bfbed0, globals=<optimized out>, locals=<optimized out>, args=<optimized out>, argcount=<optimized out>, kws=0x0, kwcount=0, defs=0x0, defcount=0, kwdefs=0x0, closure=0x0, name=<optimized out>, qualname=0x0)
    at ../Python/ceval.c:4071
#28 0x00005555556dabef in PyEval_EvalCodeEx () at ../Python/ceval.c:4092s
#29 PyEval_EvalCode (co=<optimized out>, globals=<optimized out>, locals=<optimized out>) at ../Python/ceval.c:777
#30 0x00005555557bb2d2 in run_mod (mod=<optimized out>, filename=<optimized out>, globals=0x7ffff6c34548, locals=0x7ffff6c34548, flags=<optimized out>, arena=<optimized out>) at ../Python/pythonrun.c:991
#31 0x00005555557bd5bd in PyRun_FileExFlags (fp=0x555555c68ed0, filename_str=<optimized out>, start=<optimized out>, globals=0x7ffff6c34548, locals=0x7ffff6c34548, closeit=1, flags=0x7fffffffdbcc) at ../Python/pythonrun.c:944
--Type <RET> for more, q to quit, c to continue without paging--
#32 0x00005555557bd7c4 in PyRun_SimpleFileExFlags (fp=0x555555c68ed0, filename=<optimized out>, closeit=1, flags=0x7fffffffdbcc) at ../Python/pythonrun.c:411
#33 0x000055555576ced0 in run_file (p_cf=0x7fffffffdbcc, filename=0x555555bfd390 L"greeter_client.py", fp=0x555555c68ed0) at ../Modules/main.c:328
#34 Py_Main (argc=<optimized out>, argv=<optimized out>) at ../Modules/main.c:778
#35 0x0000555555666468 in main (argc=2, argv=0x7fffffffddd8) at ../Programs/python.c:65
```

### Anything else we should know about your project / environment?
Version `1.18.0` does not exhibit this behavior.

**Comment by ericgribkoff** [comment]
https://github.com/grpc/grpc/pull/17973 is the fix for this

**Comment by srini100** [comment]
Marking this as release blocker.

**Comment by srini100** [comment]
This is good test to add for future breakages as pre-fork handlers are non-trivial.

**Comment by ericgribkoff** [comment]
Fixed by #17973. Regression testing for this will mostly be covered by https://github.com/grpc/grpc/pull/16513, but there is some value in a test like this for ensuring that fork+exec does not hang when core's fork handlers are enabled (which is the default for python builds).
